src/rule_evaluator.ts

```diff
@@ -0,0 +1,74 @@
+import { Condition, OperatorType, Rule, RuleType, AttributeValueType } from './rule';
+import { InvalidArgumentError } from './validation';
+
+export function matchesAnyRule(
+  targetingAttributes: Record<string, AttributeValueType>,
+  rules: Rule[],
+): boolean {
+  for (const rule of rules) {
+    if (matchesRule(targetingAttributes, rule)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+function matchesRule(targetingAttributes: Record<string, AttributeValueType>, rule: Rule): boolean {
+  const conditionEvaluations = evaluateRuleConditions(targetingAttributes, rule.conditions);
+  switch (rule.type) {
+    case RuleType.AND:
+      return !conditionEvaluations.includes(false);
+    case RuleType.OR:
+      return conditionEvaluations.length === 0 || conditionEvaluations.includes(true);
+  }
+}
+
+function evaluateRuleConditions(
+  targetingAttributes: Record<string, AttributeValueType>,
+  conditions: Condition[],
+): boolean[] {
+  return conditions.map((condition) => evaluateCondition(targetingAttributes, condition));
+}
+
+const NUMERIC_OPERATORS = [OperatorType.GT, OperatorType.LT, OperatorType.GTE, OperatorType.LTE];
+const STRING_OPERATORS = [OperatorType.MATCHES];
+
+function evaluateCondition(
+  targetingAttributes: Record<string, AttributeValueType>,
+  condition: Condition,
+): boolean {
+  const value = targetingAttributes[condition.attribute];
+  if (value) {
+    validateAttributeType(value, condition);
+    switch (condition.operator) {
+      case OperatorType.GTE:
+        return value >= condition.value;
+      case OperatorType.GT:
+        return value > condition.value;
+      case OperatorType.LTE:
+        return value <= condition.value;
+      case OperatorType.LT:
+        return value < condition.value;
+      case OperatorType.MATCHES:
+        return new RegExp(condition.value as string).test(value as string);
+    }
+  }
+  return false;
+}
+
+function validateAttributeType(value: any, condition: Condition) {
```

Comment:
I chose to have strict validation of the attribute value type.

There is a design consideration of whether to allow expressions like “14” > 10 (comparing a numeric string with a number) or whether to throw a validation error in this case.

Pros of strict validation:

- String comparison could work differently depending on the language. For example, “stringValue” > 10 might evaluate as TRUE in Ruby but false in Go. Also some other language might not even support the greater than argument with strings.
- It allows developers to catch errors during development if they are testing with an experiment that has defined targeting rules.

Cons of strict validation:

- The client process would receive a fatal error if the client sends bad data for the targeting attributes. This could happen if the attributes are generated dynamically, sometimes as strings, other times as numbers.
- During development, the engineer might not notice they are passing invalid data; if they are testing with an experiment that doesn’t have targeting rules, the validation would never trigger.